tensorly/decomposition/mda.py

```diff
@@ -0,0 +1,165 @@
+import numpy as np
+from .. import backend as T
+from ..base import unfold
+from ..tenalg import multi_mode_dot
+
+
+# Author: James Oldfield
+
+# License: BSD 3 clause
+
+
+def mda(X, y, ranks, n_iters=5):
+    """Multilinear Linear Discriminant Analysis (MDA).
+
+        Learns a projection matrix for each mode of the data tensor
+        to project the input tensor into a low-dimensional tensor subspace
+        where the *scatter ratio criterion* is maximised.
+
+    Parameters
+    ----------
+    X : ndarray
+        tensor data of shape (n_samples, N1, ..., NS). Note: the first dimension is the sample dimension.
+    y : ndarray
+        list of length (n_samples) containing the integer-valued class label of tensor sample i.
+    ranks : list
+        list of integers determining the dimension of the subspace for mode-k.
+    n_iters : int, optional, default is 5
+        number of steps to employ the APP scheme for.
+
+    Returns
+    -------
+    factors : list
+        list of the learnt projection matrices for each mode
+
+    Notes
+    -----
+
+    This implementation computes the *Constrained Multilinear Discriminant Analysis* (CMDA) solution as presented in [1].
+
+    Given the learnt factor matrices, one can then compute the projection along all modes with::
+
+      factors = mda(X_train, y_train, ranks, n)
+      Z = tl.tenalg.multi_mode_dot(X_train, factors, modes=[1, 2], transpose=True)
+
+    - [1] Q. Li and D. Schonfeld, "Multilinear Discriminant Analysis for Higher-Order Tensor Data Classification," in IEEE Transactions on Pattern Analysis and Machine Intelligence, vol. 36, no. 12, pp. 2524-2537, 1 Dec. 2014, doi: 10.1109/TPAMI.2014.2342214.
+
+    """
+
+    ###############
+    # check correct # of ranks have been supplied
+    ###############
+    assert len(ranks) == len(T.shape(X)[1:]), 'Expected number of ranks: {}. \
```

Comment:
It's good for end users to make the error messages as clear as possible. Something like:

```suggestion
    if len(ranks) != (T.ndim(X) - 1):
        msg = f'Got as input a tensor of shape {T.shape(X)} corresponding to {T.shape(X)[0]} samples of shape {T.shape(X)[1:]} of order {T.ndim(X) -1}.'
        msg += f'But got {len(ranks)} ranks != {T.ndim(X) -1}.'
        raise ValueError(msg)
```